proposals/0000-multiwriter.md

```diff
@@ -0,0 +1,363 @@
+
+Right now there are two "heads" of the graph: Alice's feed at seq 2, and Bob's
+feed at seq 0.
+
+Next, Alice writes a new value, and her latest entry will refer to Bob's:
+
+```
+// Feed
+
+//// Alice
+0 (/foo/bar = 'baz')
+1 (/foo/2   = '{ "some": "json" }')
+2 (''       = '')
+3 (/foo/hup = 'beep')
+
+//// Bob
+0 (/a/b     = '12')
+
+
+// Graph
+
+Alice: 0  <---  1  <---  2  <--/  3
+Bob  : 0  <-------------------/
+```
+
+Because Alice's latest feed entry refers to Bob's latest feed entry, there is
+now only one "head" in the database. That means there is enough information in
+Alice's seq=3 entry to find any other key in the database. In the last example,
+there were two heads (Alice's seq=2 and Bob's seq=0); both of which would need
+to be read internally in order to locate any key in the database.
+
+Now there is only one "head": Alice's feed at seq 3.
+
+
+## Authorization
+
+The set of hypercores are *authorized* in that the original author of the first
+hypercore in a hyperdb must explicitly denote in their append-only log that the
+public key of a new hypercore is permitted to edit the database. Any authorized
+member may authorize more members. There is no revocation or other author
+management elements currently.
+
+
+## Vector clock
+
+Each node stores a [vector clock](https://en.wikipedia.org/wiki/Vector_clock) of
+the last known sequence number from each feed it knows about. This is what forms
+the DAG structure.
+
+A vector clock on a node of, say, `[0, 2, 5]` means:
+
+- when this node was written, the largest seq # in my local fed is 0
+- when this node was written, the largest seq # in the second feed I have is 2
+- when this node was written, the largest seq # in the third feed I have is 5
+
+For example, Bob's vector clock for Alice's seq=3 entry above would be `[0, 3]`
+since he knows of her latest entry (seq=3) and his own (seq=0).
+
+The vector clock is used for correctly traversing history. This is necessary for
+the `db#heads` API as well as `db#createHistoryStream`.
+
+
+# Examples
+[examples]: #examples
+
+TODO:
+
+
+# Security and Privacy Concerns
+[privacy]: #privacy
+
+TODO:
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Significant increase in implementation, application, and user experience
+complexity.
+
+Two hard problems (merges and secure key distribution) are left to application
+developers.
+
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+TODO:
+
+- Why is this design the best in the space of possible designs?
+- What other designs have been considered and what is the rationale for not choosing them?
+- What is the impact of not doing this?
+
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+What is the technical term for the specific CRDT we are using?
+"Operation-based"?
+
+If there are conflicts resulting in ambiguity of whether a key has been deleted
+or has a new value, does `db.get(key)` return an array of `[new_value, None]`?
```

Comment:
db.get *always* return an array of `nodes` and not `values`. this is to provide more context. in case of the scenario you describe one of the nodes would have value === null (we are considering adding a .deleted flag instead)